src/java.base/share/classes/jdk/internal/util/ReferencedKeySet.java

```diff
@@ -0,0 +1,188 @@
+    }
+
+    /**
+     * Create a new {@link ReferencedKeySet} elements.
+     *
+     * @param isSoft          true if {@link SoftReference} elements are to
+     *                        be used, {@link WeakReference} otherwise.
+     * @param supplier        {@link Supplier} of the backing map
+     *
+     * @return a new set with {@link Reference} elements
+     *
+     * @param <E> the type of elements maintained by this set
+     */
+    public static <E> jdk.internal.util.ReferencedKeySet<E>
+    create(boolean isSoft, Supplier<Map<ReferenceKey<E>, ReferenceKey<E>>> supplier) {
+        return create(isSoft, false, supplier);
+    }
+
+    /**
+     * Create a new {@link ReferencedKeySet} elements.
+     *
+     * @param isSoft          true if {@link SoftReference} elements are to
+     *                        be used, {@link WeakReference} otherwise.
+     * @param useNativeQueue  true if uses NativeReferenceQueue
+     *                        otherwise use {@link ReferenceQueue}.
+     * @param supplier        {@link Supplier} of the backing map
+     *
+     * @return a new set with {@link Reference} elements
+     *
+     * @param <E> the type of elements maintained by this set
+     */
+    public static <E> ReferencedKeySet<E>
+    create(boolean isSoft, boolean useNativeQueue, Supplier<Map<ReferenceKey<E>, ReferenceKey<E>>> supplier) {
+         return new ReferencedKeySet<>(ReferencedKeyMap.create(isSoft, useNativeQueue, supplier));
+    }
+
+    /**
+     * Removes enqueued weak references from set.
+     */
+    public void removeStaleReferences() {
+        map.removeStaleReferences();
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+        return map.keySet().iterator();
+    }
+
+    @Override
+    public int size() {
+        return map.size();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return map.isEmpty();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean contains(Object o) {
+        return map.containsKey((T)o);
+    }
+
+    @Override
+    public boolean add(T e) {
+        return map.intern(e) == null;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        return map.remove(o) != null;
+    }
+
+    @Override
+    public void clear() {
+        map.clear();
+    }
+
+    /**
+     * Gets an existing element from the set, returning null if not present or
+     * the old element if previously added.
+     *
+     * @param e  element to get
+     *
+     * @return the old element if present, otherwise, null
+     */
+    public T get(T e) {
+        ReferenceKey<T> key = map.get(e);
+
+        return key == null ? null : key.get();
+    }
+
+    /**
+     * Intern an element to the set, returning the element if newly added or the
+     * old element if previously added.
+     *
+     * @param e  element to add
+     *
+     * @return the old element if present, otherwise, the new element
+     */
+    public T intern(T e) {
```

Comment:
I would like to see an additional `T intern(T e, UnaryOperator<T> internFunction)`, where the `internFunction` behaves like `LocalObjectCache::normalizeKey`.

Reply:
This was a bit tricky. Lambdas can't be used in phase1 so messed when MethodType kicked in. So I created two paths, one for normal and one for the UnaryOperator.